builder/builder.go

```diff
@@ -0,0 +1,390 @@
+package builder
+
+import (
+	"fmt"
+	"strings"
+)
+
+type Builder struct {
+	spacing string // optional spacing [s]
+	newline string
+	err     error
+
+	locals            []local
+	inputs, selectors []Expression
+	variants          []variant
+	patterns          []any
+}
+
+func New() *Builder {
+	return &Builder{
+		newline: "\n",
+		spacing: " ",
+	}
+}
+
+func (b *Builder) Build() (string, error) {
+	if b.err != nil {
+		return "", b.err
+	}
+
+	var s string
+
+	for _, v := range b.inputs {
+		s += ".input" + b.spacing + v.String() + b.newline
+	}
+
+	for _, v := range b.locals {
+		s += ".local" + b.spacing + string(v.variable) + b.spacing + "=" + b.spacing + v.expr.String() + b.newline
+	}
+
+	for _, pattern := range b.patterns {
+		s += fmt.Sprint(pattern)
+	}
+
+	if len(b.selectors) > 0 {
+		s += ".match" + b.spacing
+	}
+
+	for i, selector := range b.selectors {
+		if i == len(b.selectors)-1 { // newline after the last selector
+			s += selector.String() + b.newline
+		} else {
+			s += selector.String() + " "
+		}
+	}
+
+	for _, variant := range b.variants {
+		s += variant.String() + b.newline
+	}
+
+	return s, nil
+}
+
+func (b *Builder) String() string {
+	s, err := b.Build()
+	if err != nil {
+		panic(err)
+	}
+
+	return s
+}
+
+func (b *Builder) Newline(s string) *Builder {
+	if b.err != nil {
+		return b
+	}
+
+	b.newline = s
+
+	return b
+}
+
+func (b *Builder) Spacing(s string) *Builder {
+	if b.err != nil {
+		return b
+	}
+
+	b.spacing = s
+
+	return b
+}
+
+func (b *Builder) Text(s string) *Builder {
+	if b.err != nil {
+		return b
+	}
+
+	s = escapeSpecialChars(s)
```

Comment:
not correct

for text only three characteres should be escaped
```abnf
text-escape     = backslash ( backslash / "{" / "}" )
```

Reply:
How do you handle the case when a raw text starts (MF2 message) with `.`, e.g. `.ok.`? It should be a complex message `{{.ok.}}`.

Is it more appropriate to handle serialization in the `Build()` method?

Reply:
if mf2 string is  ".ok."
then it is simple message

Reply:
I'm afraid it is NOT.

```abnf
simple-message = [simple-start pattern]
simple-start = simple-start-char / text-escape / placeholder
simple-start-char = %x0-2D         ; omit . <--------
````

Reply:
simple message MUST start with `simple-start-char`

Reply:
i see,
then it cannot start with `.`

Reply:
it simply cannot start with `.`
neither in simpleMessage neither in ComplexMessage

so `{{.ok.}}` is also invalid mf2 syntax

Reply:
okay nevermind it can :)

Reply:
Complex message can start

```abnf
text-char = simple-start-char / "."
```